pkg/dyninst/decode/decoder_test.go

```diff
@@ -767,12 +764,81 @@ func TestDecoderPanics(t *testing.T) {
 	require.NotNil(t, stringType)
 	stringID := stringType.GetID()
 	decoder.decoderTypes[stringID] = &panicDecoderType{decoder.decoderTypes[stringID]}
-	_, err = decoder.Decode(Event{
+	_, _, err = decoder.Decode(Event{
 		Event:       output.Event(input),
 		ServiceName: "foo"},
 		&noopSymbolicator{},
-		bytes.NewBuffer(nil),
+		[]byte{},
 	)
 	require.Error(t, err)
 	require.Contains(t, err.Error(), "boom")
 }
+
+func TestDecoderFailsOnEvaluationError(t *testing.T) {
+	irProg := generateIrForProbes(t, "simple", "stringArg")
+	decoder, err := NewDecoder(irProg, &noopTypeNameResolver{})
+	require.NoError(t, err)
+	caseIdx := slices.IndexFunc(cases, func(c testCase) bool {
+		return c.probeName == "stringArg"
+	})
+	testCase := &cases[caseIdx]
+	input := testCase.eventConstructor(t, irProg)
+	var stringType ir.Type
+	for _, t := range irProg.Types {
+		if t.GetName() == "string" {
+			stringType = t
+			break
+		}
+	}
+	require.NotNil(t, stringType)
+	stringID := stringType.GetID()
+	delete(decoder.decoderTypes, stringID)
+	out, _, err := decoder.Decode(Event{
+		Event:       output.Event(input),
+		ServiceName: "foo"},
+		&noopSymbolicator{},
+		[]byte{},
+	)
+	require.NoError(t, err)
+	require.Contains(t, string(out), "no decoder type found")
+}
+
+func TestDecoderFailsOnEvaluationErrorAndRetainsPassedBuffer(t *testing.T) {
```

Comment:
Can you write a comment explaining what this test is testing? I think you should additionally assert that the prefix of the returned output also has the same prefix as buf... Also, why are you doing it twice? Maybe there's a reason but it's not self-evident, so commentary would be appreciated.